File: java/leetcode/backtracking/letter_phone.java
/**
 * 17. Letter Combinations of a Phone Number
 * Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.



Example:

Input: "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:

Although the above answer is in lexicographical order, your answer could be in any order you want.
 */

public class letter_phone {

    // backtracking is a kind of dfs
    // dfs is used for reachable problems, it stops when reaching a point.
    // backtracking is used for combination problems, it continues execution when reaching a point.

    // notes
    // mark element and next avoid next iteration of recursion repeat it
    // however, when an iteration returns, unmark the marked element. Only need to avoid one recursion call stack.

    private static final String[] KEYS = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {
        List<String> combinations = new ArrayList<>();
        if (digits == null || digits.length() == 0) {
            return combinations;
        }
        doCombination(new StringBuilder(), combinations, digits);
        return combinations;
    }

    private void doCombination(StringBuilder prefix, List<String> combinations, final String digits) {
        if (prefix.length() == digits.length()) {
            combinations.add(prefix.toString());
            return;
        }
        int curDigits = digits.charAt(prefix.length()) - '0';
        String letters = KEYS[curDigits];
        for (char c : letters.toCharArray()) {
            prefix.append(c);
            doCombination(prefix, combinations, digits);
            prefix.deleteCharAt(prefix.length() - 1);
        }
    }

}

File: golang/leetcode/backtracking/letter_phone.go
package backtracking

// 17. Letter Combinations of a Phone Number
var (
	keys = []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
)

func letterCombinations(digits string) []string {
	combinations := make([]string, 0)
	if len(digits) == 0 {
		return combinations
	}
	prefix := make([]byte, 0)
	letterTrack(prefix, &combinations, digits)
	return combinations
}

func letterTrack(prefix []byte, combinations *[]string, digits string) {
	if len(prefix) == len(digits) {
		*combinations = append(*combinations, string(prefix))
		return
	}
	var curDigits int
	curDigits = int(digits[len(prefix)] - '0')
	letters := []byte(keys[curDigits])
	for _, c := range letters {
		prefix = append(prefix, c)
		letterTrack(prefix, combinations, digits)
		prefix = prefix[:len(prefix)-1]
	}
}
